pkg/edit/edit.go

```diff
@@ -0,0 +1,108 @@
+package edit
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	errors "github.com/zgalor/weberr"
+)
+
+type machineType struct {
+	id      string
+	ccsOnly string
+	enabled string
+}
+
+type cloudRegion struct {
+	id       string
+	enabled  string
+	govcloud string
+	ccsOnly  string
+}
+
+type ConstraintMap struct {
+	entireFile   []string
+	machineTypes []machineType
+	cloudRegions []cloudRegion
+	Path         string
+}
+
+func (c *ConstraintMap) ReadConstraints(env string, appInterfaceDir string) error {
+	c.Path = path.Join(appInterfaceDir,
+		"resources/services/ocm/"+env+"/cloud-resource-constraints.configmap.yaml")
+	yamlFile, err := os.ReadFile(c.Path)
+	if err != nil {
+		return errors.Errorf("File read err: '%v'", err)
+	}
+
+	for _, line := range strings.Split(strings.TrimSuffix(string(yamlFile), "\n"), "\n") {
+		c.entireFile = append(c.entireFile, line)
+	}
+
+	return nil
+}
+
+func (c *ConstraintMap) EditConstraint(id string, ccsOnly *bool, enabled *bool, govcloud *bool) error {
+	edit := false
+	changeMade := false
+	for i, line := range c.entireFile {
```

Comment:
@hunterkepley Personally, I'd probably create a `struct` to hold the structure of the config file and unmarshall/marshall from that. I think it will make your logic easier to encapsulate and also gives you the benefit of having the structure of the file checked by the parser e.g typos in field names are caught. So I'm thinking _something_ like:

```
type MachineTypeConstraints struct {
	MachineTypes []MachineType `yaml:"instance_types"`
}

func (c *MachineType) Update(ccsOnly *string, enabled *bool) bool {
	
	changed := false
	if ccsOnly != nil {
		c.Ccs_only = ccsOnly
		changed = true
	}

	if enabled != nil {
		c.Enabled = enabled
		changed = true
	}
	
	return changed
}

# Specific function to edit MachineType constraints, `Cmd` works out which function to call
func (c *ConstraintMap) EditMachineTypeConstraint(id string, ccsOnly *bool, enabled *bool) error {
instanceTypeConstraints := InstanceTypeConstraints{}
	yamlFile, err := os.ReadFile(c.Path)
	yaml.Unmarshal(yamlFile, &instanceTypeConstraints)
	for _, c := range instanceTypeConstraints.InstanceTypes {
		if c.Id == id {
			changed := c.Update(ccsOnly, enabled)
			if !changed {
				fmt.Printf("\nNo changes were made, writing same data back into constraintmap " +
					"(use '--help' to see what you can change via flags)")
			}
		}
	}

	out, err := yaml.Marshal(instanceTypeConstraints)
	if err != nil {
		return err
	}
	err := os.WriteFile(c.path, out, 0640)
	if err != nil {
		return err
	}
	return nil
}
```

That's my personal preference. I can understand why you went the route of trying to have a single function to update the different constraint types, so please disregard this if you disagree or if there's already a reason you didn't take this approach beforehand.

Reply:
Thanks for the extra context @hunterkepley 👍